### PR TITLE
EDSC-4415: Adds spatial search drawing

### PR DESCRIPTION
**Pull request by macrouch**
# Overview

### What is the feature?

Adds spatial search drawing

### What areas of the application does this impact?

Map

# Testing

- Click each map control button, draw spatial type
- Select each spatial type from spatial dropdown by the keyword search box (except File), draw spatial type
- Manually enter spatial values and map should focus on the new spatial search
- Use advanced search to search for HUC and River Reach, when applying the search the map should focus on the new spatial search

- Polygons should be drawn with great circle arcs in geographic projection
- Circles in geographic projection should be stretched (https://en.wikipedia.org/wiki/Tissot's_indicatrix)
- Bounding boxes in polar projections should 'donut' around the poles

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by macrouch** [comment]
Simplify lines sent to CMR with reach spatial

File: static/src/js/util/map/drawSpatialSearch.js [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+            crsProjections[projectionCode]
+          )
+
+          const feature = new Feature({
+            geometry: lineInProjection
+          })
+
+          vectorSource.addFeature(feature)
+        } else {
+          // Draw a polygon for the region
```
Comment (macrouch) [inline static/src/js/util/map/drawSpatialSearch.js:77]: for the huc

File: static/src/js/util/map/styles.js [hunk 2]
```diff
@@ -84,3 +84,12 @@ export const highlightedPointStyle = (index) => new Style({
     })
   })
 })
+
+export const mbrStyle = new Style({
+  stroke: new Stroke({
+    color: '#c0392b',
+    width: 3,
+    lineDash: [2, 10],
+    opactiy: 0.8
```
Comment (macrouch) [inline static/src/js/util/map/styles.js:93]: opacity

File: tests/e2e/map/map_spatial.spec.js [hunk 3]
```diff
@@ -31,764 +52,1767 @@ test.describe.skip('Map: Spatial interactions', () => {
...
+          // Draws the spatial on the map
+          await expect(page).toHaveScreenshot('4326-point-spatial-typed.png', {
+            clip: screenshotClip
+          })
+        })
+      })
 
-        // Add the point to the map
-        await page.mouse.click(1000, 500)
+      test.describe.skip('When editing a point', () => {
```
Comment (eudoroolivares2016) [inline tests/e2e/map/map_spatial.spec.js:196]:
```suggestion
      test.describe('When editing a point', () => {
```

File: tests/e2e/map/map_spatial.spec.js [hunk 4]
```diff
@@ -31,764 +52,1767 @@ test.describe.skip('Map: Spatial interactions', () => {
...
+            clip: screenshotClip
+          })
+        })
+      })
 
-        // Enter the spatial point
-        await page.getByTestId('spatial-display_point').focus()
-        await page.keyboard.type('4.5297,42.1875')
-        await page.keyboard.up('Enter')
+      test.describe.skip('When editing a circle', () => {
```
Comment (eudoroolivares2016) [inline tests/e2e/map/map_spatial.spec.js:419]:
```suggestion
      test.describe('When editing a circle', () => {
```

File: tests/e2e/map/map_spatial.spec.js [hunk 5]
```diff
@@ -31,764 +52,1767 @@ test.describe.skip('Map: Spatial interactions', () => {
...
-              'cmr-hits': '2'
-            },
-            paramCheck: (parsedQuery) => parsedQuery?.point?.[0]?.match(/42\.\d+,-7\.\d+/)
-          },
-          {
-            body: pointBodyEdited,
-            headers: {
-              ...commonHeaders,
-              'cmr-hits': '3'
+      test.describe.skip('When editing a bounding box', () => {
```
Comment (eudoroolivares2016) [inline tests/e2e/map/map_spatial.spec.js:641]:
```suggestion
      test.describe('When editing a bounding box', () => {
```

File: tests/e2e/map/map_spatial.spec.js [hunk 6]
```diff
@@ -31,764 +52,1767 @@ test.describe.skip('Map: Spatial interactions', () => {
...
+
+          // Updates the URL
+          await expect(page).toHaveURL(/search\?sb\[0\]=42\.\d+%2C-21\.\d+%2C56\.\d+%2C-7\.\d+/)
+
+          // Populates the spatial display field
+          await expect(page.getByTestId('spatial-display_southwest-point')).toHaveValue(/-21\.\d+,42\.\d+/)
+          await expect(page.getByTestId('spatial-display_northeast-point')).toHaveValue(/-7\.\d+,56\.\d+/)
+
+          // Checking that the right number of results are loaded ensures that the route
+          // was fulfilled correctly with the succesfull paramCheck
```
Comment (eudoroolivares2016) [inline tests/e2e/map/map_spatial.spec.js:685]: nit;
```suggestion
          // was fulfilled correctly with the successful paramCheck
```
do global replace

File: tests/e2e/map/map_spatial.spec.js [hunk 7]
```diff
@@ -31,764 +52,1767 @@ test.describe.skip('Map: Spatial interactions', () => {
...
+          // Draws the spatial on the map
+          await expect(page).toHaveScreenshot('4326-box-spatial-drawn.png', {
+            clip: screenshotClip
+          })
+        })
+      })
 
-        // Add the point to the map
-        await page.mouse.click(1000, 500)
+      test.describe.skip('When editing a polygon', () => {
```
Comment (eudoroolivares2016) [inline tests/e2e/map/map_spatial.spec.js:833]:
```suggestion
      test.describe('When editing a polygon', () => {
```

**Comment by eudoroolivares2016** [comment]
I tested most of the features locally including boundary conditions on the text fields for the spatial types and didn't run into issues; this also fixes a few issues in prod

**Review by eudoroolivares2016 — APPROVED** [review]

**Review by daniel-zamora — APPROVED** [review]

File: tests/e2e/map/map_granules.spec.js [hunk 8]
```diff
@@ -236,90 +233,6 @@ test.describe('Map: Granule interactions', () => {
         })
       })
     })
-
-    test.describe.skip('When viewing OpenSearch granules with polygon spatial', () => {
```
[the remark this answers is not in the document]
Comment (macrouch) [reply tests/e2e/map/map_granules.spec.js:240]: it moved to a different file

File: static/src/js/containers/AdvancedSearchModalContainer/AdvancedSearchModalContainer.jsx [hunk 9]
```diff
@@ -101,17 +106,72 @@ const EnhancedAdvancedSearchModalContainer = withFormik({
     return advancedSearch
   },
   handleSubmit: (values, { props }) => {
+    console.log('🚀 ~ AdvancedSearchModalContainer.jsx:109 ~ values:', values)
```
[the remark this answers is not in the document]
Comment (macrouch) [reply static/src/js/containers/AdvancedSearchModalContainer/AdvancedSearchModalContainer.jsx:109]: saw it in 4417

File: static/src/js/util/map/geometryFunctions/circleGeometryFunction.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  let workingGeometry = geometry
+  if (!workingGeometry) {
+    workingGeometry = new GeometryCollection([
+      // The polygon will be what is drawn on the map
+      new Polygon([]),
+      // The center point will be saved in the GeometryCollection
+      // for use in the `drawend` event
+      new Point(coordinates[0])
+    ])
+  }
```
[the remark this answers is not in the document]
Comment (macrouch) [reply static/src/js/util/map/geometryFunctions/circleGeometryFunction.js:32]: on the first click of a new drawing

File: static/src/js/util/map/geometryFunctions/circleGeometryFunction.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  // Transform the coordinates to the geographic projection
+  const circleCenter = transform(
+    coordinates[0],
+    projection,
+    crsProjections[projections.geographic]
+  )
+  const circleLast = transform(
+    coordinates[1],
+    projection,
+    crsProjections[projections.geographic]
```
[the remark this answers is not in the document]
Comment (macrouch) [reply static/src/js/util/map/geometryFunctions/circleGeometryFunction.js:45]: we don't use this geometry function in polar projections, we just draw a circle

File: tests/e2e/map/map_spatial.spec.js (hunk 3 above)
Comment (macrouch) [reply tests/e2e/map/map_spatial.spec.js:196]: We decided not to edit spatial right now, leaving these as skipped

File: tests/e2e/map/map_spatial.spec.js (hunk 4 above)
Comment (macrouch) [reply tests/e2e/map/map_spatial.spec.js:419]: We decided not to edit spatial right now, leaving these as skipped

File: tests/e2e/map/map_spatial.spec.js (hunk 7 above)
Comment (macrouch) [reply tests/e2e/map/map_spatial.spec.js:833]: We decided not to edit spatial right now, leaving these as skipped
